pctasks/core/pctasks/core/utils/summary.py

```diff
@@ -0,0 +1,673 @@
+"""summary.py
+
+Module for generating summaries from JSON objects.
+Useful for generating Collection "summaries" and "item-assets" fields.
+"""
+
+
+from abc import abstractmethod
+from enum import Enum
+from functools import reduce
+from typing import Any, Dict, List, Optional, Set, Union, cast
+
+from pydantic import BaseModel, Field
+
+from pctasks.core.utils import map_opt
+
+
+class SummarySettings(BaseModel):
+    max_mixed_summary_samples = 4
+    max_mixed_value_samples = 4
+    max_mixed_object_list_samples = 3
+    max_mixed_key_sets = 5
+
+
+class ValueTypes(str, Enum):
+    BOOLEAN = "boolean"
+    INT = "numeric-int"
+    FLOAT = "numeric-float"
+    STRING = "string"
+    LIST = "list"
+    NULL = "null"
+
+
+class SummaryTypes(str, Enum):
+    DISTINCT = "distinct"
+    INT_RANGE = "int-range"
+    FLOAT_RANGE = "float-range"
+    OBJECT_LIST = "object-list"
+    MIXED_VALUE = "mixed-value"
+    MIXED_OBJECT_LIST = "mixed-object-list"
+    MIXED_SUMMARIES = "mixed-summaries"
+
+
+def is_value_type(v: Any) -> bool:
+    return (
+        isinstance(v, bool)
+        or isinstance(v, str)
+        or isinstance(v, float)
+        or isinstance(v, int)
+    )
+
+
+class ValueCount(BaseModel):
+    type: str
+    count: int
+
+    def add_count(self, other: "ValueCount") -> None:
+        self.count += other.count
+
+
+class BoolValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.BOOLEAN, const=True)
+    value: bool
+
+
+class IntValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.INT, const=True)
+    value: int
+
+
+class FloatValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.FLOAT, const=True)
+    value: float
+
+
+class StringValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.STRING, const=True)
+    value: str
+
+
+class NullValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.NULL, const=True)
+    value: Optional[str] = Field(default=None, const=True)
+
+
+class ListValueCount(ValueCount):
+    type: str = Field(default=ValueTypes.LIST, const=True)
+    value: List[Any]
+
+
+class PropertySummary(BaseModel):
```

Comment:
To clarify my understand here: Is the idea that you'll have many instances of these `PropertySummary` classes, and then somehow merge them together into one (or a few?). And it's possible to merge different subclass types, or am I misreading that?

If so, I've found it helpful to have a list of rules for what the expected output type is for each pair of inputs.

Reply:
There's instances for property classes for each property that's encountered. It's possible for 2 JSON docs to have the same property point to different types, so the merge can happen between types. Or for instance the merging of a DistinctValueSummary can merge with another DistinctValueSummary, hit the limit of how many distinct values are allowed, and then return instead a MixedValueSummary.

I can list out the rules. This should only be pertinent to folks developing on top of this or fixing issues, and not to pctasks users, so I'll just put a comment block in the file to explain the transformations.

Reply:
Taking back the "internal" aspect of this - that is true now, but I can see these models being used to eventually provide a service for generating collection summaries automatically, and so knowing how to interpret the output is/will be important. Will doc the different merge methods.